web/src/app/components/bar-chart/bar-chart.component.ts

```diff
@@ -0,0 +1,387 @@
+})
+
+export class BarChartComponent implements AfterViewInit {
+  /**
+   * Constants.
+   */
+  private static readonly DEFAULT_NUM_BARS: number = 30;
+  private static readonly NANO_TO_SECS: number = 10 ** 9;
+  private static readonly NANO_TO_MILLI: number = 10 ** 6;
+  private static readonly SECS_TO_HRS: number = 60 * 60;
+  private static readonly COLOR_LIGHT_GRAY: string = '#787878';
+  private static readonly COLOR_DARK_GRAY: string = '#373C38';
+  private static readonly COLOR_WHITE: string = '#eee';
+  private static readonly COLOR_WHITE_TRANS: string = '#ffffff00';
+  private static readonly DATE_FORMAT: string = 'yyyy-mm-dd hh:mm';
+  private static readonly DATE_LOCALE: string = 'en-US';
+  private static readonly COMPLETE_STATE_TAG: number = 5;
+  private static readonly BUILT_STATE_TAG: number = 14;
+  private static readonly BUILT_UNCHANGED_TAG: number = 17;
+  private static readonly SET_DEFAULT_VERSION_STATE_TAG: number = 18;
+  private static readonly ABANDONED_STATE_TAG: number = 19;
+  private static readonly STATE_TO_COLOR: { [index: number]: string } = {
+    1: '#eee',
+    3: '#2196f3',
+    4: '#d50000',
+    5: '#34a853',
+    6: '#d50000',
+    7: '#2196f3',
+    8: '#2196f3',
+    9: '#d50000',
+    10: '#2196f3',
+    12: '#d50000',
+    13: '#2196f3',
+    15: '#2196f3',
+    16: '#d50000'
+  };
+
+  /**
+   * Private variables.
+   * 
+   * The non-null assertion operator('!') suggests that there is no 
+   * issue of variables being `null` or `undefined`. Some variables are 
+   * declared as any, as their types change in different functions.
+   */
+  @ViewChild('barchart') private barChartContainer!: ElementRef; // Bar chart component
+  @Input() private pushInfos!: step189_2020.IPushInfo[] | null;
+  @Input() private currentPush!: step189_2020.IPushInfo | null;
+
+  private dataAll: Item[] = [];
+  private dataComplete: Item[] = [];
+  private totalDuration: number[] = [];
+  private completeDuration: number[] = [];
+  private svg: d3SVG | undefined;
+  private focus!: d3G;
+  private brush!: d3G;
+  private brushSelector: any;
+  private heightFocus: number = 0;
+  private heightBrush: number = 0;
+  private width: number = 0;
+  private xScaleFocus: any;
+  private yScaleFocus!: d3ScaleLinear;
+  private xScaleBrush: any;
+  private yScaleBrush!: d3ScaleLinear;
+  private xAxisFocus!: d3G;
+  private xAxisBrush!: d3G;
+  private yAxis!: d3G;
+
+  ngAfterViewInit(): void {
+    if (!this.pushInfos) { return; }
+    this.populateData(this.pushInfos);
+    this.initialChart();
+    this.updateChart(0); // Initialize the focus chart with dataAll.
+    // TODO: Call function that creates the brushSelector.
+  }
+
+  /**
+   * This function populates pushID, state, startTime and duration of given 
+   * pushes. Both arrays are in ascending order by startTime.
+   * 
+   * @param pushInfos Array for one push def
+   */
+  private populateData(pushInfos: step189_2020.IPushInfo[]): void {
```

Comment:
If this function is only updating this.dataAll we can make it return that data, make the function static and call it by doing (#28 is doing something similar):

    this.dataAll = BarChartComponent.populateData(this.pushInfos);

Reply:
It is updating `this.dataAll` and `this.dataComplete` for two bar charts and updating `totalDuration` and `completeDuration` for the future boxplot. :)

Reply:
In that case the `populateData` is not quite accurate. How about simply call this function `update` and let the comment for the function explain what is doing?